AP_ROMFS: tweak data storage to reduce size

tpwrules opened the pull request:
Some nice size reductions and code simplifications. Save roughly 300 bytes depending on board. See commit messages for details.

Tested on a CubeOrange. Made sure that it can flash its own bootloader using the button in Mission Planner and that it still boots afterward.

khancyr commented:
I think it could be a good idea, but the lost of the checksum is an issue IMHO. If we can validate the data at compile time and have the whole flash integrity check to also check the embeded data I think that would be fine . Otherwise, having a bunch of memory without integrity check will be an issue

tridge commented on Tools/ardupilotwaf/embed.py:
```diff
@@ -19,48 +19,33 @@ def embed_file(out, f, idx, embedded_name, uncompressed):
-    pad = 0
+    # ensure null termination
+    if contents[-1] != 0:
+        contents += bytes([0])
```
this adds a 0 to the contents which means for some file types we will corrupt the file (eg. fonts).
atm we ensure there is a zero at the end for consumers that treat it as text, but the zero is after the size bytes of data

tpwrules replied:
This (trailing 0 included in the size) already happened for an uncompressed ROMFS, is that a pre-existing bug?

I had trouble understanding the requirements here anyway. How does this sound:
1. The buffer returned by `find_decompress` shall have a 0 byte within it
2. If a 0 byte has to be added to make condition 1 true, then it shall be added at the end and not be reflected in the returned `size`.

I would call it verboten anyway to read past `size` bytes in the buffer, can we fix the consumers instead?

tridge replied:
> is that a pre-existing bug?

yes, it is, although the boards that use uncompressed ROMFS tend to only embed the bootloader where it doesn't matter
we could look at all uses of ROMFS and see if we can avoid the zero termination requirement

tpwrules commented:
Fixed the null termination issue. There are things that depend on it so I haven't removed it yet, though I may in the future.

Rebased also and tested again on CubeOrange and looks good.

tridge commented:
looks good, thanks!
